### PR TITLE
Display cropping tool on file entity

### PR DESCRIPTION
**Pull request by Dencev**
Here is some starting code for https://www.drupal.org/node/2617856

Removed some more testing code i have. Mostly interested why the js is not loaded on line 234?

### CODE REVIEW

File: image_widget_crop.module [hunk 1]
```diff
@@ -221,3 +221,101 @@ function image_widget_crop_library_info_alter(&$libraries, $extension) {
...
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function image_widget_crop_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
+  if (\Drupal::service('module_handler')->moduleExists('file_entity')) {
+    // We only want crop ui on images.
+    if ($form_id == 'file_image_inline_edit_form' || $form_id == 'file_image_edit_form') {
```
Comment (Berdir) [inline image_widget_crop.module]: You don't really need a module exists check if you go for a specific form_id, those only exist when that module is there. 

I'm also not yet sure if this should live here or in file_entity. Lets see what @woprrr  thinks.

File: image_widget_crop.module [hunk 2]
```diff
@@ -221,3 +221,101 @@ function image_widget_crop_library_info_alter(&$libraries, $extension) {
...
+
+      // TODO not working?
+      $from['#attached']['library'][] = 'image_widget_crop/cropper.integration';
```
Comment (Berdir) [inline image_widget_crop.module]: $from ;)

File: image_widget_crop.module (hunk 1 above)
Comment (Berdir) [inline image_widget_crop.module]: What you *do* need to check is if this actually is an image, e.g. by checking the first part of the mime type. we don't want to show this for videos or pdf's or so.

**Comment by woprrr** [comment]
It seems to me rather strange why need a hook in our case? the code that appears when file_element the process should remain in process () as image_widget done. This also will be on the media that the user seeks to upload because the widget responds only to withstand the kind of image type field.

**Comment by Berdir** [comment]
This is about the file *edit* form. Note a widget on another entity, my idea is to display it when editing the files themself. And then you need a form alter, that is correct.

This will be connected to the widget through an edit button that opens the file edit form in a dialog, similar to media/file_entity in 7.x

**Comment by Berdir** [comment]
That said, exactly because it is so different I'm not sure if this really belongs in this module. But all the UI/JS stuff is currently in this module I think, so we need to rely on it somehow..

File: image_widget_crop.module (hunk 1 above)
Comment (thenchew) [inline image_widget_crop.module]: 1. The module check you are right, probably not needed. Just the form id looks kinda generic (no module name as prefix or something) so wanted to be sure. But i will remove it.

2. The check for the image is in the $form_id. If we have an image (tested with jpg, png, gif) then its -> file_image_inline_edit_form, if its txt or pdf we get -> file_document_edit_form. That's why I didn't add an additional check.

**Comment by woprrr** [comment]
I review this and try to install file_entity into my instance to understand all this PR. ATM this PR work weel for you ? or you need more informations ?

**Comment by Dencev** [comment]
Sorry for the late response. I was sick, still a little weak. The code needs some more work (working on it today) its not saving the crop and need to address berdirs comments. So is it ok if the hook stays in image_widget_crop?

**Comment by Dencev** [comment]
Left a comment on line 234 and 237. The problem i have is getting values for all crop types that were updated. With $form_state->getValues() I don't get anything and with $form_state->getUserInput() i get the values i need but only from the last crop type. Might be because the last is overwriting the values because of the same name?

I need some help on getting the x, y, height...

**Comment by woprrr** [comment]
@Dencev I do not understand why there is a desire to re-write what already exists in terms of algorithm? The form_alter seems to be a big duplication of code in the widget and you do not seem to have any recent version of the module because there were some fix on the update positions with cropper.

Until then I have not had time to make a full return you and I will do it now. I need to know what is needed and as it currently does not work with file_entity with basic IWC? Why form_alter? Of that I would have realized I could actively help you on this.

Thank you for your valuable time !!

**Comment by Dencev** [comment]
@woprrr Sorry for wasting your time if i made the wrong approach. The goal is to display the cropping when (with file_entity) you edit the file. So i started with the form alter here.

Also more info here: https://www.drupal.org/node/2617856

**Comment by slashrsm** [comment]
As @Dencev already mentioned we're trying to bring cropping solution to file_entity edit page. We can't use widget for that and thus we took the form_alter approach. 

I agree we should try to avoid redundant code. I'd suggest to check what is duplicated and create helper function which can be used by both widget and file entity edit page.

**Comment by Berdir** [comment]
Maybe a screenshot will help.

selection_017

What you can see in the background is not the standard image widget, it's a custom widget that integrates with entity browser and allows to re-use files. It's also not an image field but a file field, so there can be videos/audios. too. And it has an edit button to edit metadata of hte file directly.

Which is why we thought that it would be a good idea to optionally support confguring cropping also in that edit form, so that we don't have to integrate it into 4+ different widgets (file_entity provides a simple one that adds an edit button and now also the entity browser integration. And there will be more widget features later, e.g. dropzonejs might also provide one)

**Comment by Berdir** [comment]
I could also fully understand if you'd prefer that this is not in image_widget_crop, since it's not at all about image_widget ;) I'm open to move it to file_entity. But it's probably better to keep it here for now, as mentioned, we might want to do some refactoring and make code re-usable. We can do that here and then decide if the form alter should be here or in file_entity.

**Comment by woprrr** [comment]
I think this is a very good idea to keep that here. And I think that is absolutely right slashrsm it is our responsibility!

"I'd suggest to check what is duplicated and create helper function can be used by qui Both widget file and edit entity page. "

I will now look at and propose something . I focus myself on the PR for the coming days.

**Comment by Dencev** [comment]
Talked with slashrsm and we are moving to https://github.com/drupal-media/image_widget_crop

**Comment by woprrr** [comment]
Yes, i purpose to create an helper function and adapt file_entity code to get iwc elements. I ve found an track to integrate easy IWC elements into file_entity form. @see the patch 
[track-test-file-form.patch.zip](https://github.com/woprrr/image_widget_crop/files/89431/track-test-file-form.patch.zip)

this prepare managed_file element like file_widget and he can be compatible with IWC i think.

What do you think @Berdir @slashrsm @Dencev

**Comment by slashrsm** [comment]
This approach seems to be tightly coupled with widgets case. I'd rather have helper code that wouldn't assume where cropper tool is displayed. This way current widget becomes just one of the implementations. 

What about custom form element for the cropper tool? This way it would be very easy to put it on any form, all logic stays in one place (element class) and it is very standard Drupal tool. 

We are doing something similar with Inline entity form. In D8 it defines form element that can live on just any form. No assumptions that would include field widget. IEF field widget then leverages from this element as it uses it to build it's own stuff. Another benefit of this approach is that it is generally easier to test. 

drupal-media/image_widget_crop is now our official repo. Let's move this discussion there.

**Pull request closed without merging**